File: multivendor_flutter_app/lib/services/vendor_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_config.dart';
import 'auth_service.dart'; 

class VendorService {
  final _authService = AuthService();

  /* ================= CREATE VENDOR ================= */
  Future<Map<String, dynamic>> createVendor(Map<String, dynamic> body) async {
    final res = await http.post(
      Uri.parse("${ApiConfig.baseUrl}/vendors"),
      headers: await _authService.headers(),
      body: jsonEncode(body),
    );

    _checkError(res);

    return jsonDecode(res.body);
  }

  /* ================= GET ALL VENDORS (ADMIN) ================= */

  Future<List<dynamic>> getAllVendors() async {
    final res = await http.get(Uri.parse('${ApiConfig.baseUrl}/vendors'), headers: await _authService.headers());

    _checkError(res);

    return jsonDecode(res.body);
  }

  /* ================= GET VENDOR BY ID ================= */

  Future<Map<String, dynamic>> getVendorById(int id) async {
    final res = await http.get(
      Uri.parse("${ApiConfig.baseUrl}/vendors/$id"),
      headers: await _authService.headers(),
    );

    _checkError(res);

    return jsonDecode(res.body);
  }

  /* ================= UPDATE VENDOR ================= */

  Future<Map<String, dynamic>> updateVendor(
    int id,
    Map<String, dynamic> body,
  ) async {
    final res = await http.put(
      Uri.parse("${ApiConfig.baseUrl}/vendors/$id"),
      headers: await _authService.headers(),
      body: jsonEncode(body),
    );

    _checkError(res);

    return jsonDecode(res.body);
  }

  /* ================= DELETE VENDOR ================= */

  Future<void> deleteVendor(int id) async {
    final res = await http.delete(
      Uri.parse("${ApiConfig.baseUrl}/vendors/$id"),
      headers: await _authService.headers(),
    );

    _checkError(res);
  }

  /* ================= GET MY VENDOR ================= */

  Future<Map<String, dynamic>> getMyVendor() async {
    final res = await http.get(
      Uri.parse("${ApiConfig.baseUrl}/vendors/me"),
      headers: await _authService.headers(),
    );

    _checkError(res);

    return jsonDecode(res.body);
  }

  /* ================= GET MY ORDER ITEMS ================= */

  Future<List<dynamic>> getMyOrderItems() async {
    final res = await http.get(
      Uri.parse("${ApiConfig.baseUrl}/vendors/me/order-items"),
      headers: await _authService.headers(),
    );

    _checkError(res);

    return jsonDecode(res.body);
  }

  /* ================= GET MY ORDERS ================= */

  Future<List<dynamic>> getMyOrders() async {
    final res = await http.get(
      Uri.parse("${ApiConfig.baseUrl}/vendors/me/orders"),
      headers: await _authService.headers(),
    );

    _checkError(res);

    return jsonDecode(res.body);
  }

  /* ================= ERROR HANDLER ================= */

  void _checkError(http.Response res) {
    if (res.statusCode >= 200 && res.statusCode < 300) return;

    try {
      final body = jsonDecode(res.body);
      throw Exception(body['message'] ?? "Server Error");
    } catch (_) {
      throw Exception("HTTP ${res.statusCode}");
    }
  }
}

File: SpringSide/src/main/java/com/abc/SpringSecurityExample/service/VendorService.java
package com.abc.SpringSecurityExample.service;

import com.abc.SpringSecurityExample.DTOs.projectDtos.*;
import com.abc.SpringSecurityExample.entity.OrderItem;
import com.abc.SpringSecurityExample.entity.Vendor;
import com.abc.SpringSecurityExample.entity.User;
import com.abc.SpringSecurityExample.mapper.OrderMapper;
import com.abc.SpringSecurityExample.mapper.VendorMapper;
import com.abc.SpringSecurityExample.repository.OrderItemRepository;
import com.abc.SpringSecurityExample.repository.UserRepository;
import com.abc.SpringSecurityExample.repository.VendorRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class  VendorService {

    private final VendorRepository vendorRepository;
    private final UserRepository userRepository;
    private final OrderItemRepository orderItemRepository;

    // ---------------- CREATE ----------------
    public VendorResponseDto createVendor(VendorRequestDto dto) {
        String userName = SecurityContextHolder.getContext().getAuthentication().getName();
        System.out.println("Logged in username:------------------------- " + userName);
        User user = userRepository.findById(userName)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (vendorRepository.existsByUser(user)) {
            throw new RuntimeException("This user already has a vendor");
        }

        Vendor vendor = VendorMapper.toEntity(dto, null);
        vendor.setUser(user);

        // TODO: Implement your SlugUtil
        String baseSlug = SlugUtil.toSlug(dto.getShopName());
        String slug = baseSlug;
        int count = 1;
        while (vendorRepository.existsBySlug(slug)) {
            slug = baseSlug + "-" + count++;
        }
        vendor.setSlug(slug);

        Vendor savedVendor = vendorRepository.save(vendor);
        return VendorMapper.toDto(savedVendor);
    }

    // ---------------- GET ALL ----------------
    public List<VendorResponseDto> getAllVendors() {
        return vendorRepository.findAll()
                .stream()
                .map(VendorMapper::toDto)
                .collect(Collectors.toList());
    }

    // ---------------- GET BY ID ----------------
    public VendorResponseDto getVendorById(Long id) {
        Vendor vendor = vendorRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Vendor not found with id: " + id));

        return VendorMapper.toDto(vendor);
    }

    // ---------------- UPDATE ----------------
    public VendorResponseDto updateVendor(Long id, VendorRequestDto dto) {
        Vendor existing = vendorRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Vendor not found with id: " + id));

        Vendor updated = VendorMapper.toEntity(dto, existing);
        Vendor savedVendor = vendorRepository.save(updated);

        return VendorMapper.toDto(savedVendor);
    }

    // ---------------- DELETE ----------------
    public void deleteVendor(Long id) {
        Vendor vendor = vendorRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Vendor not found with id: " + id));

        vendorRepository.delete(vendor);
    }

    // ---------------- GET VENDOR BY LOGGED-IN USER ----------------
    public VendorResponseDto getMyVendor() {
        Vendor vendor = getLoggedInVendor();
        return VendorMapper.toDto(vendor);
    }

    // ---------------- GET LOGGED-IN VENDOR ----------------
    public Vendor getLoggedInVendor() {
        String username = SecurityContextHolder.getContext().getAuthentication().getName();
        return vendorRepository.findByUserUserName(username)
                .orElseThrow(() -> new RuntimeException("Vendor not found"));
    }

    // ---------------- GET ALL ORDER ITEMS OF LOGGED-IN VENDOR ----------------
    public List<OrderItemResponseDTO> getMyOrderItems() {
        Vendor vendor = getLoggedInVendor();
        return orderItemRepository.findByVendor(vendor)
                .stream()
                .map(OrderMapper::mapOrderItemToDto)
                .collect(Collectors.toList());
    }


    // ---------------- GET ALL ORDERS OF LOGGED-IN VENDOR ----------------
    public List<OrderResponseDto> getMyOrders() {
        Vendor vendor = getLoggedInVendor();
        return orderItemRepository.findOrdersByVendorId(vendor.getId())
                .stream()
                .map(OrderMapper::toOrderResponseDto) // Order → OrderResponseDto
                .collect(Collectors.toList());
    }

}
